### PR TITLE
Speed-up Operation.call by avoiding unnecessary loops + more

### LINKED ISSUE #98: Progress callbacks?
Hi John,

I see that it's possible to set the `VIPS_PROGRESS` environment variable to enable some kind of feedback during long-running operations, but I'm curious whether there exists a functionality in pyvips whereby I can register a callback function similar to what appears here: https://github.com/kleisauke/net-vips/issues/31.  In addition to enabling customized display of the progress (for example, using tqdm or tqdm_notebook within Jupyter), it would presumably also enable a means to allow the user to cancel an operation like in nip2.

Many thanks in advance for any tips you can provide.

### PR DESCRIPTION
**Pull request by kleisauke**
Hi John,

I had some spare-time left to make some improvements to pyvips. It seems that the swap to new-style callbacks for `signal_connect`  does not produce any segfaults anymore. Tested both ABI and API mode with this script and [this image](https://commons.wikimedia.org/wiki/File:Mandelbrot.jpg):
<details><summary>Python script</summary>

```python
#!/usr/bin/python3

import sys
from pyvips import Image

last_percent = 0


def preeval_cb(image, progress):
    print('preeval')


def eval_cb(image, progress):
    global last_percent
    if progress.percent != last_percent:
        print('{}%, eta {}s'.format(progress.percent, progress.eta))
        last_percent = progress.percent


def posteval_cb(image, progress):
    print('posteval')


image = Image.new_from_file(sys.argv[1], access='sequential')
image.set_progress(True)
image.signal_connect('preeval', preeval_cb)
image.signal_connect('eval', eval_cb)
image.signal_connect('posteval', posteval_cb)
image.write_to_file(sys.argv[2])
```
</p>
</details>
<details><summary>Output</summary>

```bash
[kleisauke@PC-KAW ~]$ python3 signal-test.py mandelbrot.jpg x.jpg
preeval
1%, eta 0s
2%, eta 0s
3%, eta 0s
4%, eta 0s
5%, eta 0s
6%, eta 0s
7%, eta 0s
8%, eta 0s
10%, eta 26s
11%, eta 23s
12%, eta 20s
13%, eta 18s
14%, eta 18s
15%, eta 16s
16%, eta 15s
17%, eta 13s
18%, eta 12s
19%, eta 12s
20%, eta 11s
21%, eta 10s
22%, eta 10s
23%, eta 10s
24%, eta 9s
25%, eta 8s
26%, eta 8s
27%, eta 7s
28%, eta 7s
29%, eta 7s
30%, eta 6s
31%, eta 6s
32%, eta 6s
33%, eta 6s
34%, eta 5s
35%, eta 5s
36%, eta 5s
37%, eta 5s
38%, eta 4s
39%, eta 4s
40%, eta 4s
42%, eta 4s
43%, eta 3s
44%, eta 3s
45%, eta 3s
46%, eta 3s
47%, eta 3s
48%, eta 3s
49%, eta 3s
50%, eta 2s
51%, eta 2s
52%, eta 2s
53%, eta 2s
54%, eta 2s
55%, eta 2s
56%, eta 2s
57%, eta 2s
58%, eta 2s
59%, eta 2s
60%, eta 1s
61%, eta 1s
62%, eta 2s
63%, eta 2s
64%, eta 2s
65%, eta 2s
66%, eta 2s
67%, eta 1s
68%, eta 1s
69%, eta 1s
70%, eta 1s
71%, eta 1s
72%, eta 1s
74%, eta 1s
75%, eta 1s
76%, eta 1s
77%, eta 1s
78%, eta 1s
79%, eta 1s
80%, eta 0s
81%, eta 0s
82%, eta 0s
83%, eta 0s
84%, eta 0s
85%, eta 0s
86%, eta 0s
87%, eta 0s
88%, eta 0s
89%, eta 0s
90%, eta 0s
91%, eta 0s
92%, eta 0s
93%, eta 0s
94%, eta 0s
95%, eta 0s
96%, eta 0s
97%, eta 0s
98%, eta 0s
99%, eta 0s
100%, eta 0s
posteval
```
</p>
</details>

The operation speed-up change is a small micro-optimization by avoiding some unnecessary loops. This change can also be found at NetVips (commit https://github.com/kleisauke/net-vips/commit/b529231bce50973841918d324fc3ec9ff1dbb961). However, I still needs some proper benchmarks to see what the real speed increase is (I don't think more than a few microseconds).

By the way,
Do you have a test-script for the change in https://github.com/libvips/pyvips/commit/1dc365a3beb6484d609d3aa725cd871c68af2ef0? I'm interested in this to see if NetVips has the same problem and if I can reproduce it here.

This PR resolves #98.

### CODE REVIEW

File: pyvips/gvalue.py [hunk 1]
```diff
@@ -274,7 +274,7 @@ def get(self):
             # we want a ref that will last with the life of the vimage:
             # this ref is matched by the unref that's attached to finalize
             # by Image()
-            gobject_lib.g_object_ref(go)
+            gobject_lib.g_object_ref(vi)
```
Comment (kleisauke) [inline pyvips/gvalue.py]: Note that this change is somewhat unrelated to this PR. I found it while searching through the `g_object_ref` calls to see if there is another way to fix https://github.com/libvips/pyvips/commit/1dc365a3beb6484d609d3aa725cd871c68af2ef0.

**Comment by jcupitt** [comment]
Hiya, yes, I fixed the segv in callbacks, but forgot to update the issue tracking it.

I'll look through this, thanks!

File: pyvips/gvalue.py (hunk 1 above)
Comment (kleisauke) [reply pyvips/gvalue.py]: Some random thoughts: Perhaps we need to `g_object_ref` on `GValue.gobject_type` instead of `GValue.image_type` ([here](https://github.com/libvips/pyvips/blob/f5aa60a3a38ccf03d17d729821a9581c12732297/pyvips/gvalue.py#L268)).

Just like ruby-vips: https://github.com/libvips/ruby-vips/blob/117857fb1583e2e4041adfdf3e3e26d6f01848a4/lib/vips/gvalue.rb#L225-L230,

**Comment by jcupitt** [comment]
I was benchmarking Operation.call by timing the test suite, fwiw. API mode got us 20% on that, as I remember. It'd be interesting if this change makes a measurable difference.

**Comment by kleisauke** [comment]
I did some benchmarks on my server (`1x Intel® Xeon® D-1531 - 32 GB DDR4 ECC - 2x 250 GB SSD`):

```bash
[kleisauke@cdg perf]# python3 -m pyperf stats operation-call-old.json
Total duration: 12.0 sec
Start date: 2019-06-13 16:50:44
End date: 2019-06-13 16:50:59
Raw value minimum: 128 ms
Raw value maximum: 140 ms

Number of calibration run: 1
Number of run with values: 20
Total number of run: 21

Number of warmup per run: 1
Number of value per run: 3
Loop iterations per value: 1024
Total number of values: 60

Minimum:         125 us
Median +- MAD:   130 us +- 3 us
Mean +- std dev: 130 us +- 3 us
Maximum:         137 us

  0th percentile: 125 us (-4% of the mean) -- minimum
  5th percentile: 125 us (-4% of the mean)
 25th percentile: 127 us (-2% of the mean) -- Q1
 50th percentile: 130 us (-0% of the mean) -- median
 75th percentile: 133 us (+2% of the mean) -- Q3
 95th percentile: 135 us (+4% of the mean)
100th percentile: 137 us (+5% of the mean) -- maximum

Number of outlier (out of 119 us..141 us): 0
[kleisauke@cdg perf]# python3 -m pyperf stats operation-call-new.json
Total duration: 11.6 sec
Start date: 2019-06-13 16:49:02
End date: 2019-06-13 16:49:17
Raw value minimum: 124 ms
Raw value maximum: 135 ms

Number of calibration run: 1
Number of run with values: 20
Total number of run: 21

Number of warmup per run: 1
Number of value per run: 3
Loop iterations per value: 1024
Total number of values: 60

Minimum:         121 us
Median +- MAD:   126 us +- 3 us
Mean +- std dev: 126 us +- 3 us
Maximum:         131 us

  0th percentile: 121 us (-4% of the mean) -- minimum
  5th percentile: 121 us (-4% of the mean)
 25th percentile: 124 us (-2% of the mean) -- Q1
 50th percentile: 126 us (-0% of the mean) -- median
 75th percentile: 129 us (+2% of the mean) -- Q3
 95th percentile: 131 us (+4% of the mean)
100th percentile: 131 us (+4% of the mean) -- maximum

Number of outlier (out of 115 us..137 us): 0
[kleisauke@cdg perf]# python3 -m pyperf compare_to operation-call-old.json operation-call-new.json --table
+----------------+--------------------+----------------------------+
| Benchmark      | operation-call-old | operation-call-new         |
+================+====================+============================+
| Operation.call | 130 us             | 126 us: 1.03x faster (-3%) |
+----------------+--------------------+----------------------------+
```

It's just a small speed-up (1.03x faster) on a simple `Operation.call` without `kwargs` in API mode. See commit https://github.com/libvips/pyvips/pull/104/commits/d06ff91dd695d98060b1c5c42d2fcd874f4eaed7 for the relevant benchmarks.

**Comment by kleisauke** [comment]
I had a go with [Cython](https://cython.org/), see: https://github.com/kleisauke/cyvips.

On my Fedora PC, I managed to speed-up the `Operation.call` method from 70.7 us to 18.5 us (3.83x faster).

**Comment by jcupitt** [comment]
Sorry for the delay -- I simplified signal handling slightly and fixed pip8, but I think that's all.

Nice work!

**Comment by kleisauke** [comment]
Thanks for merging! I noticed that `if i not in references:` is missing here:
https://github.com/libvips/pyvips/commit/58a6d452ed92d61a737a32c661b413a82747f2c0#diff-a9fc2cb9740456112fd459f462f7343eR172

So I guess we just reintroduced the bug mentioned in https://github.com/libvips/pyvips/commit/666b65ad263148016d1ab21f1a9deaa9e0aba20b with this merge (sorry for not rebasing this PR earlier, it fell off my radar).

**Comment by kleisauke** [comment]
fwiw, I'll run some tests with the test script mentioned in https://github.com/libvips/libvips/issues/1457. Perhaps that [this change](https://github.com/libvips/pyvips/pull/104/files/f5aa60a3a38ccf03d17d729821a9581c12732297#diff-266ba5fa3a80367189d831bb452aaf55R277) could remove the reference logic [here](https://github.com/libvips/pyvips/blob/39aff939544e4dd59e1c17a1ec2227095d0992e4/pyvips/voperation.py#L165-L181), but I'm not sure.

**Comment by jcupitt** [comment]
Ooop, I put that `if not` back, nice spot. yes, it gets difficult to read old PRs with complex history :(

That reference logic is necessary because of `new_from_memory`. 

That takes a memoryview or buffer object, extracts a pointer to the underlying area of ram (and a length) and passes that to libvips to wrap with a `VipsImage`. pyvips then wraps that with an `Image` and returns the new object.

The problem is that the pyvips `Image` object contains (indirectly) a pointer to the ram represented by the memoryview. If the memoryview is GCd, we'll see a crash.

To fix this, the `Image` that `new_from_memory` returns must contain a secret Python-level reference back to the original memory object -- you can see that being recorded here:

https://github.com/libvips/pyvips/blob/master/pyvips/vimage.py#L358

But that's not enough!

Suppose we make a new `Image` that uses this result:

```python
a = pyvips.Image.new_from_memory(some_data, ...)
b = a * 10
```

Now what happens when `a` goes out of scope? libvips will keep a ref to the underlying `VipsImage`, since it knows that `b` needs it, but it does not know about the secret ref that pyvips is keeping, so that will get dropped, the memoryview will be GCed, and we'll crash again.

The solution is that in `call`, all output vips objects must inherit the set union of the refernces on all input vips objects.

I think the confusion here (for me) came because we have two "layers" of objects (the `VipsObject` layer and the Python object layer over it), each has a concept of references, and they both need to be maintained. 

If you imagine it like this:

```
a <--- b <--- c     Python Image
|      |      |
v      v      v
a' <-- b' <-- c'    C VipsImage
```

Maybe it's clearer.

**Comment by jcupitt** [comment]
I was thinking about a cleanup for `call`. It's become very complex.

How about splitting it into two parts: introspection (finding the set of inputs and outputs) and call? 

The introspection part would take an operation name (eg. `invert`) and return a structure listing all the args and flags in some form that's fast and convenient for Python. The call half would prepare the inputs, call, extract the outputs, and forward references.

It would make it easy to add an introspection cache. We could have a simple hash from operation name -> introspection data and build it on demand. It should speed up call quite a bit, and it can't get too large.

**Comment by jcupitt** [comment]
We could share the introspect method with the two doc generators and remove some code duplication too.

**Comment by jcupitt** [comment]
.. I'll have a quick go at a PR, I have some time now.

**Comment by kleisauke** [comment]
Thanks for the explanation regarding the reference logic for `new_from_memory`, I'll check if NetVips has the same problem.

Indeed, the `call` method is becoming complex. Splitting up into an introspection part and a call part is a good idea.

By the way, I noticed that there is a `@_add_doc` decorator which is called before every operation:
https://github.com/libvips/pyvips/blob/1be032667a1478e3846d691d14ade5fe2d63e6c5/pyvips/vimage.py#L126-L128
https://github.com/libvips/pyvips/blob/1be032667a1478e3846d691d14ade5fe2d63e6c5/pyvips/vimage.py#L816-L818

Doesn't these calls affect performance? It will eventually call `vips_object_get_args`, etc.

**Comment by kleisauke** [comment]
fwiw, I came across this SO post: https://stackoverflow.com/a/28059785/10952119, it seems that special `__slots__` attribute provides faster access to attributes and may reduce the memory footprint.

I had a quick go at it:
<details><summary>__slots__ patch</summary>

```diff
From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Sat, 16 Nov 2019 15:50:00 +0100
Subject: [PATCH] Use the special __slots__ attribute

For faster attribute access and to reduce the memory footprint.

diff --git a/pyvips/gobject.py b/pyvips/gobject.py
index 1111111..2222222 100644
--- a/pyvips/gobject.py
+++ b/pyvips/gobject.py
@@ -41,6 +41,7 @@ class GObject(object):
     """Manage GObject lifetime.
 
     """
+    __slots__ = '_handles', 'pointer', 'gobject'
 
     def __init__(self, pointer):
         """Wrap around a pointer.
diff --git a/pyvips/gvalue.py b/pyvips/gvalue.py
index 1111111..2222222 100644
--- a/pyvips/gvalue.py
+++ b/pyvips/gvalue.py
@@ -29,6 +29,7 @@ class GValue(object):
     GValue lifetime is managed automatically.
 
     """
+    __slots__ = 'pointer', 'gvalue'
 
     # look up some common gtypes at init for speed
     gbool_type = type_from_name('gboolean')
diff --git a/pyvips/vimage.py b/pyvips/vimage.py
index 1111111..2222222 100644
--- a/pyvips/vimage.py
+++ b/pyvips/vimage.py
@@ -83,16 +83,23 @@ def _run_cmplx(fn, image):
 
 
 # https://stackoverflow.com/a/22409540/1480019
-def _with_metaclass(mcls):
-    def decorator(cls):
-        body = vars(cls).copy()
-        # clean out class body
-        body.pop('__dict__', None)
-        body.pop('__weakref__', None)
-
-        return mcls(cls.__name__, cls.__bases__, body)
-
-    return decorator
+# https://github.com/benjaminp/six/blob/33b584b2c551548021adb92a028ceaf892deb5be/six.py#L846-L861
+def _with_metaclass(metaclass):
+    """Class decorator for creating a class with a metaclass."""
+    def wrapper(cls):
+        orig_vars = cls.__dict__.copy()
+        slots = orig_vars.get('__slots__')
+        if slots is not None:
+            if isinstance(slots, str):
+                slots = [slots]
+            for slots_var in slots:
+                orig_vars.pop(slots_var)
+        orig_vars.pop('__dict__', None)
+        orig_vars.pop('__weakref__', None)
+        if hasattr(cls, '__qualname__'):
+            orig_vars['__qualname__'] = cls.__qualname__
+        return metaclass(cls.__name__, cls.__bases__, orig_vars)
+    return wrapper
 
 
 # decorator to set docstring
@@ -135,6 +142,7 @@ class Image(pyvips.VipsObject):
     """Wrap a VipsImage object.
 
     """
+    __slots__ = '_references',
 
     # private static
 
diff --git a/pyvips/voperation.py b/pyvips/voperation.py
index 1111111..2222222 100644
--- a/pyvips/voperation.py
+++ b/pyvips/voperation.py
@@ -44,6 +44,7 @@ class Operation(pyvips.VipsObject):
     This class wraps the libvips VipsOperation class.
 
     """
+    __slots__ = 'object',
 
     # cache nickname -> docstring here
     _docstring_cache = {}
```
</details>

I'll try some benchmarks now.

**Comment by jcupitt** [comment]
The `_add_doc` decorators update `_docstring_cache`, so it'll only make the string on the first call (hopefully).

**Comment by jcupitt** [comment]
Oh huh, I'd not come across `__slots__`, interesting.

**Comment by jcupitt** [comment]
I've made a PR with a separate Introspect class. `call` is about 30% faster on your nice perf test, and it's a bit smaller too.

**Comment by jcupitt** [comment]
Up to 50% now!
